File: python/faceAveraging.py
import os
import sys
import cv2
import dlib
import numpy as np
import faceBlendCommon as fbc

# 读取指定的文件夹下的所有的jpg图片
def readImagesPath(path):
    """
    从指定的文件夹下面读取所有的jpg图片
    参数:
    ============
    path:文件夹路径
    """
    pass

    imagePaths=[]
    for filePath in sorted(os.listdir(path)):
        fileExt = os.path.splitext(filePath)[1]
        if fileExt in [".jpg", ".jpeg"]:
            print(filePath)
            imagePaths.append(os.path.join(path, filePath))
    
    return imagePaths


# 入口
if __name__ == "__main__":
    # model
    PREDICTOR_PATH = "../common/resources/shape_predictor_68_face_landmarks.dat"

    # face detector
    faceDetector = dlib.get_frontal_face_detector()
    landmarkDetector = dlib.shape_predictor(PREDICTOR_PATH)

    dirName = "../data/images/girls"

    # 读取所有的图片
    imagePaths = readImagesPath(dirName)

    if len(imagePaths) == 0:
        print("没有读取到对应的图片")
        sys.exit(0)
    
    images = []
    allPoints = []

    for imagePath in imagePaths:
        im = cv2.imread(imagePath)
        if im is None:
            print("image:{} 读取失败".format(imagePath))
        else:
            print("image:{} 读取成功".format(imagePath))
            points=fbc.getLandmarks(faceDetector, landmarkDetector, im, 2)
            if len(points) > 0:
                allPoints.append(points)
                im = np.float32(im)/255.0
                images.append(im)
            else:
                print("未检测到landmarks")
    
    if len(images) == 0:
        print("没有检测到合适的图像")
        sys.exit(0)
    
    # 定义输出图像的尺寸
    w = 300
    h = 300

    boundaryPts = fbc.getEightBoundaryPoints(h, w)
    
    numImages = len(imagePaths)
    numLandmarks = len(allPoints[0])

    # 归一化图片和特征点
    imagesNorm = []
    pointsNorm = []

    # 平均值
    pointsAvg = np.zeros((numLandmarks,2), dtype=np.float32)

    # 将图片规整到输出坐标系中并求解平均值
    for i, img in enumerate(images):
        points = allPoints[i]

        points = np.array(points)

        # print(points)

        img, points = fbc.normalizeImagesAndLandmarks((h,w),img,points)

        cv2.imshow("{}".format(i), img)

        pointsAvg = pointsAvg + (points/(1.0*numImages))

        # 添加边界点
        points = np.concatenate((points, boundaryPts), axis=0)

        pointsNorm.append(points)
        imagesNorm.append(img)

    pointsAvg = np.concatenate((pointsAvg, boundaryPts), axis=0)
    
    # 计算delaunay三角形
    rect = (0, 0, w, h)
    dt = fbc.calculateDelaunayTriangles(rect, pointsAvg)
    
    # 输出图像
    output = np.zeros((h, w, 3), dtype=np.float)
    
    for i in range(0, numImages):
        imWarp = fbc.warpImage(imagesNorm[i], pointsNorm[i], pointsAvg.tolist(), dt)

        output = output + imWarp
    
    output = output/(1.0*numImages)
    cv2.imshow("Result", output)
    cv2.waitKey(0)
    

File: cpp/faceAveraging.cpp
#include "faceBlendCommon.hpp"
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <algorithm>
#include <vector>

// 在linux或者Unix系统中dirent.h文件并不是包含的，但是在Windows上是与包含的，所以我们分开指定
#ifdef _WIN32
#include "dirent.h"
#elif __APPLE__
#include "TargetConditionals.h"
#if TARGET_OS_MAC
#include <dirent.h>
#else
#error "非Mac系统"
#endif
#elif __linux__
#include <dirent.h>
#elif __unix__
#include <dirent.h>
#else
#error "我真的不知道你的系统了"
#endif

void readFileName(string dirName, vector<string> &imageFnames)
{
    DIR *dir;
    struct dirent *ent;
    int count = 0;

    string imgExt = "jpg";
    vector<string> files;

    if ((dir = opendir(dirName.c_str())) != NULL)
    {
        // 打印出所有文件的名字
        while ((ent = readdir(dir)) != NULL)
        {
            if (strcmp(ent->d_name, ".") == 0 || strcmp(ent->d_name, "..") == 0)
            {
                continue;
            }
            string temp_name = ent->d_name;
            cout << "文件名:" << temp_name << endl;
            files.push_back(temp_name);
        }
        std::sort(files.begin(), files.end());
        for (int i = 0; i < files.size(); i++)
        {
            string path = dirName;
            string fname = files[i];

            if (fname.find(imgExt, (fname.length() - imgExt.length())) != std::string::npos)
            {

                path.append(fname);
                cout << "文件路径:" << path << endl;
                imageFnames.push_back(path);
            }
            else
            {
                cout << "未找到指定类型的文件" << endl;
            }
        }
        closedir(dir);
    }
}

int main(int argc, char const *argv[])
{
    // 面部检测器
    dlib::frontal_face_detector faceDetector = dlib::get_frontal_face_detector();
    // 形状检测器
    dlib::shape_predictor landmarkDetector;

    // 初始化形状检测器
    dlib::deserialize("../common/resources/shape_predictor_68_face_landmarks.dat") >> landmarkDetector;

    // 读取所有的图片
    string dirName = "../data/images/girls";
    if (!dirName.empty() && dirName.back() != '/')
    {
        dirName += '/';
    }

    // 读取目录下的图片文件
    vector<string> imageNames, ptsNames;
    readFileName(dirName, imageNames);

    // 如果没有文件的话可以退出
    if (imageNames.empty())
    {
        cout << "没有制定类型的文件!" << endl;
        return EXIT_FAILURE;
    }

    // 获取所有图片的脸部关键点
    vector<vector<Point2f>> allPoints;
    vector<Mat> images;
    for (size_t i = 0; i < imageNames.size(); i++)
    {
        Mat img = imread(imageNames[i]);
        if (!img.data)
        {
            cout << "图片:" << imageNames[i] << "读取失败!" << endl;
        }
        else
        {
            vector<Point2f> points = getLandmarks(faceDetector, landmarkDetector, img, 1.0);
            if (points.size() > 0)
            {
                allPoints.push_back(points);
                img.convertTo(img, CV_32FC3, 1 / 255.0);
                images.push_back(img);
            }
        }
    }

    if (images.empty())
    {
        cout << "没有检测到有脸的图像" << endl;
        return EXIT_FAILURE;
    }

    int numImages = images.size();

    // 归一化的图片和点
    vector<Mat> imagesNorm;
    vector<vector<Point2f>> pointsNorm;

    // 平局值之后的关键特征点
    vector<Point2f> pointsAvg(allPoints[0].size());

    // 定出输出图像的尺寸
    Size size(600, 600);

    // 定义进行Delaunay剖分的8个边界点
    vector<Point2f> boundaryPts;
    getEightBoundaryPoints(size, boundaryPts);

    for (size_t i = 0; i < images.size(); i++)
    {
        vector<Point2f> points = allPoints[i];

        Mat img_temp;
        normalizeImagesAndLandmarks(size, images[i], img_temp, points, points);

        // 计算标记点的平均位置
        for (size_t k = 0; k < points.size(); k++)
        {
            pointsAvg[k] += points[k] * (1.0 / numImages);
        }

        // 将边界点加入进去
        for (size_t m = 0; m < boundaryPts.size(); m++)
        {
            points.push_back(boundaryPts[m]);
        }

        pointsNorm.push_back(points);
        imagesNorm.push_back(img_temp);
    }

    // 边界点
    for (size_t m = 0; m < boundaryPts.size(); m++)
    {
        pointsAvg.push_back(boundaryPts[m]);
    }

    // Delaunay矩形
    Rect rect(0, 0,size.width, size.height);
    vector<vector<int>> dt;
    calculateDelaunayTriangles(rect, pointsAvg, dt);

    Mat output = Mat::zeros(size, CV_32FC3);

    for(size_t i=0; i<numImages; i++)
    {
        Mat img;
        warpImage(imagesNorm[i], img, pointsNorm[i], pointsAvg, dt);

        output = output + img;

        ostringstream win;
        win << "Result"<< i;

        imshow(win.str(), output);
    }

    output = output / (double)numImages;

    // 将合成图像保存下来
    imwrite("../data/images/result.jpg", output*255);

    imshow("Result", output);
    waitKey(0);

    return 0;
}
